pypeit/core/datacube.py

```diff
@@ -267,88 +313,102 @@ def extract_standard_spec(stdcube, subpixel=20):
+        xx, yy = np.meshgrid(x, y, indexing='ij')
+        # Generate a Gaussian kernel
+        intflux = 1
+        xo, yo = popt[1], popt[2]
+        fwhm2sigma = 1.0 / (2 * np.sqrt(2 * np.log(2)))
+        sigma_x, sigma_y = optfwhm*fwhm2sigma, optfwhm*fwhm2sigma
+        theta, offset, = 0.0, 0.0
+        optkern = gaussian2D((xx, yy), intflux, xo, yo, sigma_x, sigma_y, theta, offset).reshape(wl_img.shape)
+        # Normalise the kernel
+        optkern /= np.sum(optkern)
+
+    optkern_masked = optkern * mask[:,:,0]
+    # Normalise the white light image
+    optkern_masked /= np.sum(optkern_masked)
+    asrt = np.argsort(optkern_masked, axis=None)
+    # Need to ensure that the number of pixels in the object profile is even
+    if asrt.size % 2 != 0:
+        # Remove the pixel with the lowest kernel weight.
+        # It should be a zero value (given the mask), so it doesn't matter if we remove it
+        asrt = asrt[1:]
+    # Now sort the indices of the pixels in the object profile
+    tmp = asrt.reshape((asrt.size//2, 2))
+    objprof_idx = np.append(tmp[:,0], tmp[::-1,1])
+    objprof = optkern_masked[np.unravel_index(objprof_idx, optkern.shape)]
+
+    # Now slice the datacube and inverse variance cube into a 2D array
+    spat, spec = np.meshgrid(objprof_idx, np.arange(numwave), indexing='ij')
+    spatspl = np.apply_along_axis(np.unravel_index, 1, spat, optkern.shape)
+    # Now slice the datacube and corresponding cubes/vectors into a series of 2D arrays
+    numspat = objprof_idx.size
+    flxslice = (spatspl[:,0,:], spatspl[:,1,:], spec)
+    flxcube2d = _flxcube[flxslice].T
+    ivarcube2d = _ivarcube[flxslice].T
+    gpmcube2d = np.logical_not(bpmcube[flxslice].T)
+    waveimg = wave.reshape((numwave,1)).repeat(numspat, axis=1)
+    skyimg = np.zeros((numwave, numspat))  # Note, the residual sky has already been subtracted off _flxcube
+    oprof = objprof.reshape((1,numspat)).repeat(numwave, axis=0)
+    thismask = np.ones_like(flxcube2d, dtype=bool)
+
+    # Now do the optimal extraction
+    extract.extract_optimal(flxcube2d, ivarcube2d, gpmcube2d, waveimg, skyimg, thismask, oprof,
+                            sobj, min_frac_use=0.05, fwhmimg=None, base_var=None, count_scale=None, noise_floor=None)
+
+    # TODO :: The optimal extraction may suffer from residual DAR correction issues. This is because the
+    #      :: object profile assumes that the white light image represents the true spatial profile of the
+    #      :: object. One possibility is to fit a (linear?) model to the ratio of box/optimal extraction
+    #      :: and then apply this model to the optimal extraction. This is a bit of a fudge.
+    # Note that extract.extract_optimal() stores the optimal extraction in the
+    # sobj.OPT_COUNTS, sobj.OPT_COUNTS_SIG, and sobj.OPT_COUNTS_IVAR attributes.
+    # We need to store the fluxed extraction into the FLAM attributes (a slight fudge).
+    if fluxed:
```

Comment:
I agree, it would be good to have a consistent approach, but I can't see a way to do this sensibly. I'll have a think about it some more and report back.